File: templates/javascript/v1/designtemplates/memento/textEditor.js
const EMPTY_BUFFER = "";

export class TextSnapshot {
  constructor(state) {
    this.state = state;
  }
}

export class TextEditor {
  constructor() {
    this.buffer = EMPTY_BUFFER;
  }

  append(text) {
    this.buffer += text;
  }

  current() {
    return this.buffer;
  }

  save() {
    return new TextSnapshot(this.current());
  }

  restore(snapshot) {
    this.buffer = snapshot.state;
  }
}

File: templates/java/v1/designtemplates/memento/TextEditor.java
package designtemplates.memento;

public class TextEditor {
    private final StringBuilder buffer = new StringBuilder();

    public void append(String text) {
        buffer.append(text);
    }

    public String current() {
        return buffer.toString();
    }

    public TextSnapshot save() {
        return new TextSnapshot(current());
    }

    public void restore(TextSnapshot snapshot) {
        buffer.setLength(0);
        buffer.append(snapshot.getState());
    }
}
